Port to meson

oreo639 opened the pull request:
If there is not interest in this, feel free to let me know.

I decided to do this as libtool/autoconf was giving me a lot of trouble with cross compiling and there isn't really a good way to do both native and cross compiling due to how the dependencies work.
This makes cross compiling much easier as meson supports an `exe_wrapper` command which can be set to, for example, qemu.

I verified that the same files get outputted with both this and autoconf (except for the .la files which shouldn't be installed anyway, and the .a files, which can be enabled in meson with `-Ddefault_library=both` if so desired. meson by default uses the equivalent of `--disable-static`):
https://gist.github.com/oreo639/cd4d3c53d8bc4d90203d647debb53dbd
(confls is autoconf and mesonls is meson)

I also verified that the resulting package can be installed and works correctly (and can be used to build ibus-anthy and fcitx-anthy).

I didn't implement `update_params`/`update_params0`/`update_params2` since I'm not sure how you want to handle those. ([fujiwarat/anthy-unicode/calctrans/Makefile.am](https://github.com/fujiwarat/anthy-unicode/blob/360c112cff27fe82fff07bb255a48e40ab7a0849/calctrans/Makefile.am#L16-L39))
Let me know if I missed anything.
I didn't add meson to the autoconf dist, if that is something you want me to do, let me know. (meson dist works by pulling the files from git so it should be fine on that side)

I also added a fix for crashing when `getpwuid()` returns null when cross compiling (due to `add_val("HOME", pw->pw_dir);` trying to dereference a null ptr). I made it such that if `getpwuid()` returns null, HOME doesn't get added to the dictionary. I'm not sure how desirable this is. diclib should still be able to fallback on `getenv()` when that is the case, however if you have other suggestions, let me know.

Of course, let me know if I missed something.

oreo639 committed:
Port to meson

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

oreo639 committed:
src-diclib: fix crash when getpwuid() returns null

This fixes a null pointer dereference when cross-compiling with an exe_wrapper
script when the passwd cannot be used.

anthy_conf_get_str() will still fallback to getenv().

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

fujiwarat commented:
Thank you very much for your patches. let me review it later.

fujiwarat commented:
Seems it requires `emacs` command but `/usr/share/aclocal-1.17/lispdir.m4` can set  `am_cv_lispdir='${datadir}/emacs/site-lisp'`.
I wish to keep source files of *.c in "Port to meson" patch.

oreo639 commented:
> Seems it requires `emacs` command but `/usr/share/aclocal-1.17/lispdir.m4` can set `am_cv_lispdir='${datadir}/emacs/site-lisp'`.

It should only require emacs if `-Demcas=...` is specified, although the logic can be simplified greatly (e.g. making `-Demacs` a `feature` option and adding `-Demacs_path=...` instead of checking for `-Demacs=no`) which I will do.

> I wish to keep source files of *.c in "Port to meson" patch.

I assume you mean to pull the modifications of source files in to separate commits?
(Regarding the changes in the "Port to meson" patch, meson always runs commands from the root of the build direcotry so input/output file paths need to be able to be specified from cli args for them to be able to be outputted in the correct place)

fujiwarat commented:
I assume -Demacs_path= and -Demacs= are "auto" or "" by default and sets the default values.
https://src.fedoraproject.org/rpms/anthy-unicode/blob/rawhide/f/anthy-unicode.spec

> meson always runs commands from the root of the build direcotry so input/output file paths need to be able to be specified from cli args for them to be able to be outputted in the correct place

Thank you. I will check them again but currently I think it would be better to separate those parts.

oreo639 committed:
calctrans: allow specifying output directory from cli args

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

oreo639 committed:
depgraph: allow specifying output file from cli args

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

oreo639 committed:
mkanthydic: allow specifying output file from cli args

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

oreo639 committed:
mkworddic: allow specifying output file from cli args

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

fujiwarat commented:
If I run `meson setup builddir --prefix=/usr`, `anthy-unicode.conf` includes `"${prefix}"`.

oreo639 commented:
> If I run `meson setup builddir --prefix=/usr`, `anthy-unicode.conf` includes `"${prefix}"`.

I did that because that is how it worked with autotools, do you want me to change that?

fujiwarat commented:
> I did that because that is how it worked with autotools, do you want me to change that?

Sorry, it is generated by configure but not Makefile. OK, you don't have to change it.

fujiwarat commented:
I made `anthy-unicode-1.0.0.20240207.tar.xz` with `meson dist` and run `rpmbuild -bb `[anthy-unicode.spec](https://fujiwara.fedorapeople.org/ibus/anthy-unicode/anthy-unicode.spec) in Fedora and then, `anthy-unicode.conf` includes `"${prefix}"`.

It's not reproduced with [anthy-unicode-1.0.0.20240502.tar.gz](https://github.com/fujiwarat/anthy-unicode/releases/download/1.0.0.20240502/anthy-unicode-1.0.0.20240502.tar.gz) and `rpmbuild`, which was generated with autotool in Fedora because the Fedora's configure option is `--prefix=/usr --datadir=/usr/share` so the `$datadir` does not include `$prefix`.

I think this would be a regression.

oreo639 commented:
> It's not reproduced with [anthy-unicode-1.0.0.20240502.tar.gz](https://github.com/fujiwarat/anthy-unicode/releases/download/1.0.0.20240502/anthy-unicode-1.0.0.20240502.tar.gz) and `rpmbuild`, which was generated with autotool in Fedora because the Fedora's configure option is `--prefix=/usr --datadir=/usr/share` so the `$datadir` does not include `$prefix`.
> 
> I think this would be a regression.

In this case it is just a difference in how meson works.

`meson setup build --prefix=/usr --datadir=/usr/share` would still result in `get_option('datadir')` returning `share`, not `/usr/share`. (You can confirm this using `message(get_option('datadir'))`)

Afaik there isn't a way to distinguish `--prefix=/usr --datadir=/usr/share` and `--prefix=/usr --datadir=share`, as they both output the same with meson, which is different from autoconf.

If I configure it to e.g. `--datadir=/opt/share` then I still get the following as expected:
```
#
prefix /usr
datadir /opt/share
ANTHYDIR /opt/share/anthy-unicode/
DIC_FILE /opt/share/anthy-unicode/anthy.dic
ZIPDICT_EUC /opt/share/anthy-unicode/zipcode.t
WORDS_FILE /usr/share/dict/words
# for compilation and testing
INDEPWORD indepword.txt
DEPWORD master.depword
```

So it shouldn't cause any regressions afaik.

oreo639 commented:
I can make it so it is `get_option('prefix') / get_option('datadir')` instead of `'${prefix}' / get_option('datadir')` (which would make it always be the full path).
I didn't do that since that would change the output from autoconf, but its fine either way.

You can also use meson to generate the pkgconfig file without needing a `*.pc.in` file, I didn't use that for the same reason (that would do `-I${includedir}/anthy-unicode-1` with `includedir=${prefix}/include` instead of `-I${includedir}` with `includedir=${prefix}/include/anthy-unicode-1`):
https://mesonbuild.com/Pkgconfig-module.html

fujiwarat committed:
meson: Use str.format()

Also replace ${prefix} with get_option('prefix').

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

fujiwarat commented:
Sorry for the late response.
I added a small patch to refactor the branch.
1. Replace `replace ${prefix}` with `get_option('prefix')` while I'm not sure why only `get_option('datadir')` does not work
2. Replace the version '1.0.0.20240207' with the date command. I tried to have a conditional version with `configuration_data().set('VERSION'...)` at first but it does not work when I create the tarball with `meson dist`. I also thought the conditional version in `project()` with `get_option()` and `run_command()` but probably it's not a meson design. I added a fixed date suffix at the moment.
3. Added `BUG=` line in each your patch.

It's an interesting idea to use meson to generate the pkgconfig file without needing a *.pc.in file but my understanding is `libanthy_unicode` in anthy-unicode/src-main/meson.build and `libdiclib` in anthy-unicode/src-diclib/meson.build needs to be moved to anthy-unicode/meson.build to get the library flags and I keep the `configuration_data()` way at the moment.

oreo639 commented:
Thanks for looking at it. Overall, the changes look good to me.

> 2\. Replace the version '1.0.0.20240207' with the date command. ...

The main reason I didn't do that is because the version wouldn't be baked in when doing meson dist.
There are two ways to handle that though:
1. Determine if it is a tarball and decide which version to use based on that: https://gitlab.freedesktop.org/accountsservice/accountsservice/-/blob/main/generate-version.sh?ref_type=heads https://gitlab.freedesktop.org/accountsservice/accountsservice/-/blob/main/meson.build?ref_type=heads#L3
2. Use [`meson.add_dist_script()`](https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonadd_dist_script) to write the date to a file and use the date from that file if it exists. (and just call date if it doesn't): https://github.com/GNOME/phodav/blob/45453d2f6ec5cb9acb0e05b3bf93e78b2eb8e754/meson.build#L12.

Edit: I did a basic implementation of the second approach which seems to work: https://github.com/oreo639/anthy-unicode/commit/6d0e1edd632eef81c1eeebc78868e2f3a0112103

fujiwarat commented:
Thank you for the review.

> 1. Determine if it is a tarball and decide which version to use based on that...

Right, I have noticed it and wrote it; conditional version in `project()` with `get_option()`  and `run_command()`.

> 2. Use [meson.add_dist_script()](https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonadd_dist_script) to write the date to a file and use the date from that file if it exists.

Thank you. I have noticed `meson.add_dist_script()` but I didn't get the usage.
But seems that way does not work; https://github.com/fujiwarat/anthy-unicode/commit/9398efc3b43c5c1fa1763a0a0a9bf810e5c033cc

`meson dist` command still outputs meson-dist/anthy-unicode-1.0.0.tar.xz .

Actually I also tried `meson rewrite kwargs set project / version '1.2.3'` and it does not work.

I think the project data is saved in meson-private/build.dat and I gave up the conditional version.

fujiwarat commented:
> Edit: I did a basic implementation of the second approach which seems to work: https://github.com/oreo639/anthy-unicode/commit/6d0e1edd632eef81c1eeebc78868e2f3a0112103

I missed it. will try it.

fujiwarat commented:
> Edit: I did a basic implementation of the second approach which seems to work: https://github.com/oreo639/anthy-unicode/commit/6d0e1edd632eef81c1eeebc78868e2f3a0112103

It also does not work?

```
+anthy_released = false
+if not anthy_released
+meson.add_dist_script('sh', '-c',
+                      'echo @0@>"$MESON_DIST_ROOT/.datestamp"'.format(
+                      meson.project_version().split('.')[-1]))
+
+endif
```

I think the `anthy_released` does not switch the condition.

fujiwarat commented:
Added `meson_dist.py` to follow the automake way.

oreo639 commented:
> I think the `anthy_released` does not switch the condition.

Maybe I am misunderstanding, but, for that to work with `-Danthy_released`, you would need to do `anthy_released = get_option('anthy_released')` with anthy_released defined in `meson_options.txt`.

fujiwarat commented:
> but, for that to work with -Danthy_released

Right, it's actually the first option; I have noticed it and wrote it; conditional version in `project()` with `get_option()` and `run_command()`. However I think that option is a public parameter and users can change the behavior against my intention.
I conclude the conditional version is not in the meson design at present unfortunately.

fujiwarat committed:
meson: Add meson_dist.py to update ChangeLog and NEWS files

You could run the following commands:
git tag 1.0.0.`date '+%Y%m%d'`
meson setup builddir
cd builddir
meson dist

BUG=https://github.com/fujiwarat/anthy-unicode/pull/12

fujiwarat commented:
Thank you very much for your patch. The next release '1.0.0.20260127'  will be a last version before the official version.
It would be great if you also could contribute https://github.com/fujiwarat/kasumi-unicode .

fujiwarat commented:
Unfortunately there is a regression #14
